### ISSUE TITLE
Custom lines not always rendered

### ISSUE THREAD

**Issue by Bernhard-Hiller** [osmandapp/OsmAnd-tools#328]
During the last years, I got a lot of experience with creating maps for Garmin. But the device is starting to rot, so I switched to an Android device, and tried OsmAnd. I'd like to change some features of the default style.
As a first - and likely easy - step, I'd like to add elevation contour lines. I used a Geofabrik extract of Chile, with added contour lines, and cut the area from 73-74W and 39-40S from it.
In the rendering_types.xml file (dowloaded today), I added at the end (just before </osmad_types>):

	<category name="elevation">
		<type tag="contour_ext" value="elevation_major" minzoom="10" nameTags="ele" />
        <type tag="contour_ext" value="elevation_medium" minzoom="11" nameTags="ele" />
        <type tag="contour_ext" value="elevation_minor" minzoom="13" nameTags="ele" />
    </category>

Furthermore, I created a file BERNIE.render.xml with

    <order>
		<filter layer="1" tag="contour" value="elevation" order="87"/>
		<filter tag="contour" value="elevation" order="57"/>
	</order>

and

    <line>
		<switch>
			<case tag="contour_ext" value="elevation_major" strokeWidth="2" color="#000000" />
			<case tag="contour_ext" value="elevation_medium" strokeWidth="2" color="#000000" />
			<case tag="contour_ext" value="elevation_minor" strokeWidth="1" color="#000000" />
		</switch>
	</line>

Then I started the map creation with the user interface of OsmAndMapCreator. It created an output file of about 5 MB.

The map is shown in OsmAnd, and the new contour lines are normally also shown.
But not always. I created a screenshot, available at Google Photos:
https://photos.app.goo.gl/UtbTiZU5QKp8eLvf8

As you can see, north-east of Niebla, the contour lines are missing. 
But the screenshot of JOSM (same location) provides proof that the contour lines are present.

How does that happen?

**Comment by vshcherb** [comment]
Did you try to convert obf with obf2osm inspector.sh to check whether map was created correctly?
Did you have same objects probably enabled? Please be sure that you used negative ids for elevation lines

**Comment by Bernhard-Hiller** [comment]
Thanks for mentioning the "inspector" script. I did not know about it.
Its command line output seems to be OK:

C:\Program Files (x86)\OpenStreetMap\OsmAndMapCreator>java.exe -Djava.util.logging.config.file=logging.properties -Xms64
M -Xmx512M -cp "OsmAndMapCreator.jar;lib/*.jar" net.osmand.obf.BinaryInspector -osm=C:\Users\Bernie\osmand\Surchico.osm
c:\Users\Bernie\osmand\Surchico.obf
Sep 26, 2019 9:26:07 AM net.osmand.osm.MapRenderingTypes initFromInputStream
INFORMATION: Time to init poi types 63
Binary index Surchico.obf version = 2 edition = Thu Sep 26 09:23:25 CEST 2019
1 Map data Surchico - 4308278 bytes
        1.1 Map level minZoom = 15, maxZoom = 22, size = 2432055 bytes
                Bounds (left top - right bottom) : -73.9949, -39 NE - -73, -40 NE
        1.2 Map level minZoom = 13, maxZoom = 14, size = 1194191 bytes
                Bounds (left top - right bottom) : -73.9949, -39 NE - -73, -40 NE
        1.3 Map level minZoom = 11, maxZoom = 12, size = 610966 bytes
                Bounds (left top - right bottom) : -73.9949, -39 NE - -73, -40 NE
        1.4 Map level minZoom = 9, maxZoom = 10, size = 59234 bytes
                Bounds (left top - right bottom) : -73.9949, -39 NE - -73.0001, -40 NE
2 Routing data Surchico - 413035 bytes
        Bounds (left top - right bottom) : -73.7005, -39.0001 NE - -73, -40 NE
3 Address data Surchico - 397028 bytes
        3.1 Address part size=289405 bytes
        3.3 Address part size=23056 bytes
        3.2 Address part size=1806 bytes
4 Poi data Surchico - 435441 bytes
5 Transport data Surchico - 80 bytes
        Bounds (left top - right bottom) : -180, 85.0511 NE - -180, 85.0511 NE

But it produces an empty output file.
Only 

C:\Program Files (x86)\OpenStreetMap\OsmAndMapCreator>inspector.bat -vaddress -vcitynames -vstreetgroups -vstreets -vbuildings -vintersections -vmap -vstats -vmapobjects -vmapcoordinates -osm=C:\Users\Bernie\osmand\Surchico.osm -vpoi -vrouting -vtransport c:\Users\Bernie\osmand\Surchico.obf

produced an osm file of some 7 MB  (after tons of output to the command line).
When trying to open it in Josm, Josm complains:

Die Datei 'Surchico.osm' konnte nicht gelesen werden. 
Fehlermeldung:
Zeile 4, Spalte 6: Verarbeitungsanweisungsziel, das "[xX][mM][lL]" entspricht, ist nicht zulässig.

Well, the first lines of that file are:

<?xml version='1.0' encoding='UTF-8'?>
<osm version='0.6'>
</osm >
<?xml version='1.0' encoding='UTF-8'?>
<osm version='0.6'>
	<node id = '1' version='1' lat='-39.997242' lon='-73.694145' >
	</node >

As we can immediately see, there is a duplication of the xml header in line 4, which is obviously invalid (i.e. that's a bug in inspector). Also the osm tag is duplicated. I edited that with notepad++.

Now, the file can be opened and displayed in Josm. 
No lines corresponding to the elevation contours are shown.
But neither are lines corresponding to coast_line, riverbank etc. shown.

Are some special further parameters required to get those data into the osm file, too? Or is that a further bug of the inspector utility?

**Comment by Bernhard-Hiller** [comment]
Edit to above post (the "Edit" link does not work with my browser ...)
After "Well, the first lines of that file are:", I pasted a few lines of xml which do net get displayed here.
Let me try to strap some of the "less than" and "greater than" characters to get it visible:

?xml version='1.0' encoding='UTF-8'?
osm version='0.6'
/osm 
?xml version='1.0' encoding='UTF-8'?
osm version='0.6'
	node id = '1' version='1' lat='-39.997242' lon='-73.694145' 
	/node

**Comment by Bernhard-Hiller** [comment]
I do not understand what you mean with "Did you have same objects probably enabled?".

As with the negative ids for the elevation lines, I am very astonished. To my knowledge, ids of existing items are always positive. Negative ids are only used by Josm for indicating "new" ids which are not yet committed to the database.
I used srtm2osm (see https://wiki.openstreetmap.org/wiki/Srtm2Osm ) for the creation of the elevation lines. That program starts with the maximum value of a signed 64bit integer as the first node/way id by default - which cannot be used due to the fact that that number is a magic value in splitter/mkgmap (by the way, "splitter" could be an interesting tool in the tool chain for generating maps of large areas). The ids I used are very large ("9023372036854775807"), and they work well with the splitter/mkgmap tool chain, and can be displayed correctly in Josm.

**Comment by Bernhard-Hiller** [comment]
After sending the output of inspector into a text file, I can see that the elevation lines are present:

OBJECT BY TYPE STATS: 
contour-elevation                   [   3699]     1000 KB: coord [ 293431]      928 KB, names       12 KB  
contour_ext-elevation_minor         [   2816]      775 KB: coord [ 227593]      720 KB, names        9 KB  
contour_ext-elevation_medium        [    532]      198 KB: coord [  59941]      188 KB, names        2 KB  
contour_ext-elevation_major         [    351]       26 KB: coord [   5897]       19 KB, names        0 KB  

But I did not count them in the original pbf file.

**Comment by Bernhard-Hiller** [comment]
Just counted them in Josm. The numbers are always less:

contour_ext elevation_minor 2591
contour_ext elevation_medium 464
contour_ext elevation_major 346

For the creation of the contourlines, I used theaparameter "-maxwaynodes 1000", i.e. a "way" must not contain more than 1000 nodes. 
My impression is that OsmAndMapCreator splits "large" ways into smaller ways, and the limit could be something like 320 nodes. Is that correct?

**Comment by vshcherb** [comment]
You need to be specific with params what you get in osm file i.e. you might want
```-vmap -vmapobjects -vmapcoordinates -osm=.... ``` that's it
optionally -zoom=, -bbox=

**Comment by Bernhard-Hiller** [comment]
Thanks for the parameters. Now I tried 
inspector.bat -vmap -vmapobjects -vmapcoordinates -osm=C:\Users\Bernie\osmand\Surchico.osm c:\Users\Bernie\osmand\Surchico.obf
which created a 56 MB (instead of 7 MB only) osm file. It contains all the contour lines. 
They received additional tags:
ele_limit (value e.g. 500)
original_id (extremely negative value, e.g. -1841616395753535)
osmand_id (even more negative, e.g. -235726898656452373)
Beyond that, the encoding of non-ASCII characters is wrong.
"Punta Loncoy�n" should be "Punta Loncoyén"

But, anyway, this means that the data are present in the obf file. So I guess that the problem might be located in the rendering process of OsmAnd, not in the map creation process.

**Comment by Bernhard-Hiller** [comment]
Next step to total confusion: since I know the inspector tool now, I created the obf file with the original rendering_types.xml. And ... the elevation contour lines are in the file. And completely missing in OsmAnd.

**Comment by Bernhard-Hiller** [comment]
I did further checks, and eventually did a tag_transform on all relevant values, such that "native" rendering should not interfere. The effect is still the same.
Hence, I think that it might rather be an issue of the rendering engine in the OsmAnd app than an issue in the tools for map creation, but I cannot exclude that either.
Hence I opened a ticket for OsmAnd app: https://github.com/osmandapp/Osmand/issues/7615
